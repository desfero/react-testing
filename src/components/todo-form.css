.filter {
    margin: 10px 0;
    display: flex;
    align-items: center;
}

.filter span {
    margin: 0 5px;
}