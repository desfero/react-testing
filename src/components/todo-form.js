import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { changeFilter } from '../actions';

import './todo-form.css';

export class TodoFormBase extends Component {
    static propTypes = {
        handleSubmit: PropTypes.func.isRequired
    };

    state = {title: ''};

    filterOptions = ['ALL', 'COMPLETED', 'UNCOMPLETED'];

    handleChange = event => {
        this.setState({title: event.target.value});
    };

    handleKeyUp = event => {
        if (event.keyCode === 27) this.clearInput();
    };

    handleSubmit = event => {
        event.preventDefault();
        const title = this.state.title.trim();

        if (title.length) {
            this.props.handleSubmit(title);
        }

        this.clearInput();
    };

    clearInput() {
        this.setState({title: ''});
    }

    render() {
        return (
            <Fragment>
                <form onSubmit={this.handleSubmit} noValidate>
                    <input
                      autoComplete="off"
                      autoFocus
                      className="form-control"
                      maxLength="64"
                      onChange={this.handleChange}
                      onKeyUp={this.handleKeyUp}
                      placeholder="What needs to be done?"
                      type="text"
                      value={this.state.title}
                    />
                </form>
                <div className="filter">
                    <span>Filter: </span>
                    <select className="custom-select" value={this.filter} onChange={e => this.props.changeFilter(e.target.value)}>
                        {this.filterOptions.map(filter => <option value={filter} key={filter}>{filter}</option>)}
                    </select>
                </div>
            </Fragment>
        );
    }
}

function mapStateToProps(state) {
    return {
        filter: state.filter
    }
}

const mapDispatchToProps = {
    changeFilter
};

export const TodoForm = connect(mapStateToProps, mapDispatchToProps)(TodoFormBase);
