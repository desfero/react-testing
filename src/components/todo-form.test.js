import React from 'react';
import { shallow } from 'enzyme';
import { TodoFormBase } from './todo-form';

describe('todo-form', () => {
    it('should correctly render component', () => {
        const handleSubmit = () => {};
        const wrapper = shallow(<TodoFormBase handleSubmit={handleSubmit}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should submit handle input change', () => {
        jest.mock('./todo-form');
        const handleSubmit = () => {};
        const wrapper = shallow(<TodoFormBase handleSubmit={handleSubmit}/>);

        const input = wrapper.find('input');
        const value = 'new todo';
        input.value = value;
        input.simulate('change');

        expect(TodoFormBase.handleChange).toHaveBeenCalled();

        jest.unmock('./todo-form');
    });

    // it('should submit the form', () => {
    //     const handleSubmit = jest.fn();
    //     const wrapper = shallow(<TodoFormBase handleSubmit={handleSubmit}/>);
    // });
});