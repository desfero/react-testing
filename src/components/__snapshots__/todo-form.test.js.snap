// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-form should correctly render component 1`] = `
<React.Fragment>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      autoFocus={true}
      className="form-control"
      maxLength="64"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="What needs to be done?"
      type="text"
      value=""
    />
  </form>
  <div
    className="filter"
  >
    <span>
      Filter: 
    </span>
    <select
      className="custom-select"
      onChange={[Function]}
    >
      <option
        key="ALL"
        value="ALL"
      >
        ALL
      </option>
      <option
        key="COMPLETED"
        value="COMPLETED"
      >
        COMPLETED
      </option>
      <option
        key="UNCOMPLETED"
        value="UNCOMPLETED"
      >
        UNCOMPLETED
      </option>
    </select>
  </div>
</React.Fragment>
`;
